File: node/services/dynamodb_service.js
import { 
  DynamoDBClient, 
  PutItemCommand, 
  UpdateItemCommand,
  DeleteItemCommand, 
  QueryCommand,
  ListTablesCommand,
  DeleteTableCommand,
  CreateTableCommand 
} from "@aws-sdk/client-dynamodb";
import { getRandomInt, makeSampleJobPayload } from '../common/utils.js'
import Schedule from '../common/schedule.js'
import { v4 as uuidv4 } from 'uuid';

class DynamoDBService {
  constructor(metrics){
    this.maxShardId = 10
    this.queryLimit = 5
    this.tableName = 'deali_schedules'
    this.client = new DynamoDBClient({
      endpoint: 'http://localhost:8000'
    });
    this.metrics = metrics
  }

  makePutItemCommand(tableName, schedule){
    const putItemRequest = {
      TableName: tableName,
      Item: {
        shard_id: {S: schedule.shardId},
        date_token: {S: schedule.dateToken},
        job_status: {S: schedule.jobStatus},
        job_spec: {S: JSON.stringify(schedule.jobSpec)}
      }
    }
    return new PutItemCommand(putItemRequest)
  }

  makeOverdueJobsQueryCommand(partition, timestamp, jobStatus){
    const input = {
      TableName: this.tableName,
      KeyConditionExpression: 'shard_id = :shardId and date_token < :dateToken',
      FilterExpression: 'job_status = :jobStatus',
      ExpressionAttributeValues: {
        ':shardId': {S: partition.toString()},
        ':dateToken' : {S: timestamp.toString()},
        ':jobStatus' : {S: jobStatus}
      },
      Limit: this.queryLimit
    }
    return new QueryCommand(input)
  }

  makeUpdateItemCommand(schedule, oldStatus, newStatus){
    const updateItemRequest = {
      TableName: this.tableName,
      Key: {
        shard_id: {S: schedule.shardId},
        date_token: {S: schedule.dateToken}
      },
      ConditionExpression: 'job_status = :oldStatus',
      ExpressionAttributeValues: {
        ':oldStatus': {S: oldStatus},
        ':newStatus': {S: newStatus},
      },
      UpdateExpression: 'SET job_status = :newStatus',
      ReturnValues: 'UPDATED_NEW'
    }
    return new UpdateItemCommand(updateItemRequest);
  }

  makeDeleteItemCommand(schedule){
    const deleteItemRequest = {
      TableName: this.tableName,
      Key: {
        shard_id: {S: schedule.shardId},
        date_token: {S: schedule.dateToken}
      },
      ConditionExpression: 'job_status = :acquired',
      ExpressionAttributeValues: {
        ':acquired': {S: 'ACQUIRED'}
      }
    }
    return new DeleteItemCommand(deleteItemRequest)
  }

  decodeSchedule(ddbItem){
    const shardId = ddbItem.shard_id.S
    const dateToken = ddbItem.date_token.S
    const jobStatus = ddbItem.job_status.S
    const jobSpec = JSON.parse(ddbItem.job_spec.S)
    return new Schedule(shardId, dateToken, jobStatus, jobSpec)
  }

  // 임시 - 테스트를 위한 random item put
  async addJob(tableName){
    const sampleSchedule = new Schedule(
      getRandomInt(this.maxShardId),
      `${Date.now()}#${uuidv4()}`,
      'SCHEDULED',
      makeSampleJobPayload()
    )
    const command = this.makePutItemCommand(tableName, sampleSchedule);
    return this.client
    .send(command)
    .then(()=>sampleSchedule);
  }

  async getOverdueJobs(partition){
    const start = Date.now()
    const command = this.makeOverdueJobsQueryCommand(partition, Date.now(), 'SCHEDULED');
    return this.client
    .send(command)
    .then((response)=>{
      this.metrics.getOverdueJobs(Date.now()-start, partition)
      return {
        schedules: response.Items.map(this.decodeSchedule),
        shouldImmediatelyQueryAgain: response.Count == this.queryLimit || !!response.LastEvaluatedKey
      }
    })
  }

  async updateStatus(schedule, oldStatus, newStatus){
    const command = this.makeUpdateItemCommand(schedule, oldStatus, newStatus)
    return this.client
    .send(command)
    .then((response)=>{
      const updatedStatus = response.Attributes.job_status.S
      schedule.jobStatus = updatedStatus
      return schedule
    })
  }

  async deleteDispatchedJob(schedule){
    const command = this.makeDeleteItemCommand(schedule)
    return this.client
    .send(command)
  }

  async isTableExists(tableName){
    const command = new ListTablesCommand({})
    const response = await this.client.send(command)
    return response['TableNames'].includes(tableName)
  }

  async deleteTable(tableName){
    const command = new DeleteTableCommand({TableName: tableName})
    return this.client.send(command)
  }

  async createTable(tableName){
    const command = new CreateTableCommand({
      TableName: tableName,
      AttributeDefinitions: [
        {
          AttributeName: 'shard_id',
          AttributeType: 'S'
        },
        {
          AttributeName: 'date_token',
          AttributeType: 'S'
        }
      ],
      KeySchema: [
        {
          AttributeName: 'shard_id',
          KeyType: 'HASH'
        },
        {
          AttributeName: 'date_token',
          KeyType: 'RANGE'
        }
      ],
      BillingMode: 'PAY_PER_REQUEST'
    })
    return this.client.send(command)
  }
}

export default DynamoDBService

File: golang/services/dynamodb_service.go
package ddbsvc

import (
	"fmt"
	"strconv"
	"time"

	"github.com/aws/aws-sdk-go/aws"
	"github.com/aws/aws-sdk-go/aws/session"
	"github.com/aws/aws-sdk-go/service/dynamodb"
	"github.com/aws/aws-sdk-go/service/dynamodb/dynamodbattribute"
	"github.com/aws/aws-sdk-go/service/dynamodb/expression"
)

var dynamo *dynamodb.DynamoDB

const TABLE_NAME = "deali_schedules"
const QUERY_LIMIT = 5

type ScheduleQueryResponse struct {
	Schedules                   []Schedule
	ShouldImmediatelyQueryAgian bool
}

type Schedule struct {
	ShardId   string `json:"shard_id"`
	DateToken string `json:"date_token"`
	JobStatus string `json:"job_status"`
	JobSpec   string `json:"job_spec"`
}

// These init() functions can be used within a package block
// and regardless of how many times that package is imported,
// the init() function will only be called once.
func init() {
	fmt.Println("!!!! init ddb service !!!!!")
	dynamo = connectDynamo()
}

func connectDynamo() *dynamodb.DynamoDB {
	sess := session.Must(session.NewSession(&aws.Config{
		Region:   aws.String("ap-northeast-2"),
		Endpoint: aws.String("http://localhost:8000"),
	}))
	svc := dynamodb.New(sess)
	return svc
}

func makeOverdueQueryExpression(partition int, timestamp int64, jobStatus string) *dynamodb.QueryInput {
	keyCond1 := expression.Key("shard_id").Equal(expression.Value(strconv.Itoa(partition)))
	keyCond2 := expression.Key("date_token").LessThan(expression.Value(strconv.FormatInt(timestamp, 10)))
	filt := expression.Name("job_status").Equal(expression.Value(jobStatus))
	expr, _ := expression.NewBuilder().
		WithKeyCondition(keyCond1.And(keyCond2)).
		WithFilter(filt).
		Build()

	input := &dynamodb.QueryInput{
		TableName:                 aws.String(TABLE_NAME),
		KeyConditionExpression:    expr.KeyCondition(),
		FilterExpression:          expr.Filter(),
		ExpressionAttributeNames:  expr.Names(),
		ExpressionAttributeValues: expr.Values(),
		Limit:                     aws.Int64(QUERY_LIMIT),
	}
	return input
}

func makeUpdateItemInput(schedule Schedule, oldStatus string, newStatus string) *dynamodb.UpdateItemInput {
	update := expression.Set(expression.Name("job_status"), expression.Value(newStatus))
	cond := expression.Name("job_status").Equal(expression.Value(oldStatus))
	expr, _ := expression.NewBuilder().WithCondition(cond).WithUpdate(update).Build()

	input := &dynamodb.UpdateItemInput{
		TableName: aws.String(TABLE_NAME),
		Key: map[string]*dynamodb.AttributeValue{
			"shard_id": {
				S: aws.String(schedule.ShardId),
			},
			"date_token": {
				S: aws.String(schedule.DateToken),
			},
		},
		ConditionExpression:       expr.Condition(),
		ExpressionAttributeNames:  expr.Names(),
		ExpressionAttributeValues: expr.Values(),
		UpdateExpression:          expr.Update(),
		ReturnValues:              aws.String("ALL_NEW"),
	}
	return input
}

func makeDeleteItemInput(schedule Schedule) *dynamodb.DeleteItemInput {
	cond := expression.Name("job_status").Equal(expression.Value("ACQUIRED"))
	expr, _ := expression.NewBuilder().WithCondition(cond).Build()
	input := &dynamodb.DeleteItemInput{
		TableName: aws.String(TABLE_NAME),
		Key: map[string]*dynamodb.AttributeValue{
			"shard_id": {
				S: aws.String(schedule.ShardId),
			},
			"date_token": {
				S: aws.String(schedule.DateToken),
			},
		},
		ExpressionAttributeNames:  expr.Names(),
		ExpressionAttributeValues: expr.Values(),
		ConditionExpression:       expr.Condition(),
	}
	return input
}

func GetOverdueJobs(partition int) ScheduleQueryResponse {
	now := time.Now().UnixMilli()
	input := makeOverdueQueryExpression(partition, now, "SCHEDULED")
	res, err := dynamo.Query(input)
	if err != nil {
		fmt.Println(err)
	}

	var schedules []Schedule
	dynamodbattribute.UnmarshalListOfMaps(res.Items, &schedules)
	return ScheduleQueryResponse{
		Schedules:                   schedules,
		ShouldImmediatelyQueryAgian: *res.Count == QUERY_LIMIT || res.LastEvaluatedKey != nil,
	}
}

func UpdateStatus(schedule Schedule, oldStatus string, newStatus string) Schedule {
	input := makeUpdateItemInput(schedule, oldStatus, newStatus)
	res, err := dynamo.UpdateItem(input)
	if err != nil {
		fmt.Println(err)
	}
	dynamodbattribute.UnmarshalMap(res.Attributes, &schedule)
	return schedule
}

func DeleteDispatchedJob(schedule Schedule) {
	input := makeDeleteItemInput(schedule)
	_, err := dynamo.DeleteItem(input)
	if err != nil {
		fmt.Println(err)
	}
}
